hw15-structural_patterns/src/test/java/ru/otus/handler/ComplexProcessorTest.java

```diff
@@ -0,0 +1,167 @@
+
+        //when
+        var result = complexProcessor.handle(message);
+
+        //then
+        verify(processor1, times(1)).process(eq(message));
+        verify(processor2, times(1)).process(eq(message));
+        assertThat(result).isEqualTo(message);
+    }
+
+    @Test
+    @DisplayName("Тестируем обработку исключения")
+    void handleExceptionTest() {
+        //given
+        var message = new Message.Builder(1L).field8("field8").build();
+
+        var processor1 = mock(Processor.class);
+        when(processor1.process(eq(message))).thenThrow(new RuntimeException("Test Exception"));
+
+        var processor2 = mock(Processor.class);
+        when(processor2.process(eq(message))).thenReturn(message);
+
+        var processors = List.of(processor1, processor2);
+
+        var complexProcessor = new ComplexProcessor(processors, (ex) -> {
+            throw new TestException(ex.getMessage());
+        });
+
+        //when
+        assertThatExceptionOfType(TestException.class).isThrownBy(() -> complexProcessor.handle(message));
+
+        //then
+        verify(processor1, times(1)).process(eq(message));
+        verify(processor2, never()).process(eq(message));
+    }
+
+    @Test
+    @DisplayName("Тестируем уведомления")
+    void notifyTest() {
+        //given
+        var message = new Message.Builder(1L).field9("field9").build();
+
+        var listener = mock(Listener.class);
+
+        var complexProcessor = new ComplexProcessor(new ArrayList<>(), (ex) -> {
+        });
+
+        complexProcessor.addListener(listener);
+
+        //when
+        complexProcessor.handle(message);
+        complexProcessor.removeListener(listener);
+        complexProcessor.handle(message);
+
+        //then
+        verify(listener, times(1)).onUpdated(eq(message), eq(message));
+    }
+
+    // todo: 3. Сделать процессор, который будет выбрасывать исключение в четную секунду (сделайте тест с гарантированным результатом)
+    @Test
+    @DisplayName("Тестируем процессор, который выбрасывает исключение в четную секунду")
+    void evenSecondExceptionProcessorTest() {
+        //given
+        var message = new Message.Builder(1L).build();
+
+        var processor =
+                spy(new ProcessorThrowExceptionEvenSecond(
+                        ()-> LocalDateTime.now().withSecond(2)));
+
+        var complexProcessor = new ComplexProcessor(Collections.singletonList(processor), (ex) -> {
+            throw new TestException(ex.getMessage());
+        });
+
+        //when
+        assertThatExceptionOfType(TestException.class).isThrownBy(() -> complexProcessor.handle(message));
+
+        //then
+        verify(processor, times(1)).process(eq(message));
+    }
+
+    // todo: 4. Сделать Listener для ведения истории: старое сообщение - новое (подумайте, как сделать, чтобы сообщения не портились)
+    @Test
+    @DisplayName("Тестируем ведение истории, проверяем, что сообщения не портятся")
+    void HistoryListenerMessageMutationTest() {
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.